numpy/_core/_exceptions.py

```diff
@@ -38,26 +38,30 @@ def __init__(self, ufunc):
 @_display_as_base
 class _UFuncNoLoopError(UFuncTypeError):
     """ Thrown when a ufunc loop cannot be found """
-    def __init__(self, ufunc, dtypes):
+    def __init__(self, ufunc, dtypes, ambiguous_promoter):
         super().__init__(ufunc)
         self.dtypes = tuple(dtypes)
+        self.ambiguous_promoter = ambiguous_promoter
 
     def __str__(self):
-        return (
+        err = (
             "ufunc {!r} did not contain a loop with signature matching types "
-            "{!r} -> {!r}"
+            "{!r} -> {!r}."
         ).format(
             self.ufunc.__name__,
             _unpack_tuple(self.dtypes[:self.ufunc.nin]),
             _unpack_tuple(self.dtypes[self.ufunc.nin:])
         )
+        if self.ambiguous_promoter:
+            err += " (ambiguous promoters found)"
```

Comment:
I think the one disadvantage is that this error message no longer makes clear that this is an internal bug, so should be reported. Maybe include the `NOTE: This indicates...` from the RuntimeError? Also, is this actionable? I think the user could work around it by passing a signature argument. If so, the error message would ideally say that.

Reply:
Now that we’re using generic promoters in numpy on purpose with this known limitation, is it an internal bug that needs to be reported? I don’t think it is.

Can you explain how I could use `signature` to work around this error? For the stringdtype example, there’s no sensible answer for multiplying two strings and it wouldn’t make sense to ever do that. Maybe there’s another situation where one of the promoters does make sense?

Reply:
I indeed presumed, maybe incorrectly, that if the promotion was ambiguous, at least one of the possible promoters was actually reasonable. In that case, actively selecting something should work, no?

Reply:
Well, the best solution is to have the `AnyInt` dtype.  You could add a `(String, String, String)` promoter (removing the ambiguity) which just returns `(String, String, String)` and then the current machinery will decide to give up searching (correctly).

But not sure that hack is worthwhile, since we know the better solution.  (As said, I am OK with mutating the error type either way.)

Reply:
I tried adding the `(String, String, String)` promoter and still got the ambiguous loop error, I didn’t look closely at why that didn’t work though.

Reply:
OK, that surprises me.  Since that clearly matches best and it should be picked and I am very sure returning the DTypes unmodified just aborts the search.  But maybe no need to dive in right now.

Reply:
Ah-ha! It's because the order you add the promoters matters. If I add the `(String, String, String)` promoter before the other two, it works.

Reply:
That sounds like a bug - is the search perhaps given up once an ambiguity is detected, instead of trying to find a better one?

Reply:
Ah :(, but a bit annoying to support, as you need to compare to multiple candidates.

Reply:
I guess one would need to keep track of whether the current best candidate is ambiguous, but otherwise keep on going.

Reply:
You would have to keep track of all the best candidates (plural) and hope that at the end there is only one left.

Reply:
Not necessarily, still just need to keep track of the best, but also have a flag, which you set if an equal one appears, and clear if a better one comes by. If at the end, the flag is set, you return an error.

Reply:
Hm, yes, in the general case my scheme may indeed not work.... So perhaps the real solution is `AbsInt`...